File: aho_corasick.rb
#!/usr/bin/ruby

module AhoCorasick
	def self.transitions_table(pattern, alphabet)
		l = alphabet.length # tamanho do alfabeto
		n = pattern.length
		tT = Hash.new
		alphabet.each do |char|
			tT[[0, char]] = 0
		end
		n.times do |j|
			c = pattern[j]
			prev = tT[[j, c]]
			tT[[j,c]] = j + 1
			alphabet.each do |char|
				tT[[j+1, char]] = tT[[prev, char]]
			end
		end

		tT
	end

	def self.match_sring(string, pattern, alphabet)
		tT = transitions_table(pattern, alphabet)
		s = 0
		n = string.length
		m = pattern.length
		occ = []
		n.times do |i|
			s = tT[[s, string[i]]]
			occ << (i - m + 1) if s == m
		end

		occ
	end
end

a = AhoCorasick.transitions_table("aba", ['a', 'b'])
# a = AhoCorasick.match_sring("abracadabra", "abr", ['a','b','c','d','r'])
puts a.inspect

File: AhoCorasick.java
import java.util.Vector;

public class AhoCorasick {
  public static Integer[][] transitionTable(String pattern) {
    Integer m = pattern.length();
    Integer n = 256; // tamanho do alfabeto
    Integer[][] tT = new Integer[m][n];
    for(int i = 0; i < n; i++) { // preenche primeira linha da tabela
      tT[0][i] = 0;
    }
    for(int j = 0; j < m; j++) {
      Integer c = (int) pattern.charAt(j);
      Integer prev = tT[j][c];
      tT[j][c] = j + 1;
      for(int k = 0; k < n && j < m -1; k++) {
        tT[j+1][k] = tT[prev][k];
      }
    }

    return tT;
  }

  public static Object[] matchString(String string, String pattern) {
    Integer[][] tT = transitionTable(pattern);
    Integer s = 0;
    Integer n = string.length();
    Integer m = pattern.length();
    Vector<Integer> occ = new Vector<Integer>();
    for(int i = 0; i < n; i++) {
      s = tT[s][(int) string.charAt(i)];
      if(s == m) {
        occ.add(i - m + 1);
        s = 0;
      }
    }

    return occ.toArray();
  }

  public static void main(String[] args) {
    // Integer[][] table = transitionTable("aba");
    // printTable(table);
    Object[] occ = matchString("abracadabra", "abr");
    for(int i = 0; i < occ.length; i++) {
      System.out.println(occ[i]);
    }
  }

  public static void printTable(Integer[][] table) {
    for(int i = 0; i < table.length; i++) {
      System.out.print(i + ": ");
      for(int j = 0; j < table[0].length; j++) {
        Integer s = table[i][j];
        if(s != 0) {
          System.out.print(" " + (char) j + ", " + table[i][j] + " ");
        }
      }
      System.out.println();
    }
  }
}